src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// This struct is used to read over a memeory stream in parts, in order to avoid reading the entire stream into memory.
+    /// It functions as a wrapper around <see cref="Utf8JsonStreamReader"/>, while maintaining a stream and a buffer to read from.
+    /// must call <see cref="CompleteRead"/> to return the buffer to the pool when completed.
+    /// </summary>
+    internal ref struct Utf8JsonStreamReader
+    {
+        private static readonly char[] DelimitedStringDelimiters = new char[] { ' ', ',' };
+
+        // The UT8 BOM characters which may appear at the start of a JSON file.
+        private ReadOnlySpan<byte> _utf8Bom = new byte[] { 0xEF, 0xBB, 0xBF };
+        private Utf8JsonReader _reader;
+        // The buffer is used to read from the stream in chunks.
+        private byte[] _buffer;
+        // The stream is the source of the JSON data.
+        private Stream _stream;
+        private bool _complete;
+
+        internal Utf8JsonStreamReader(Stream stream)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _complete = false;
+            _stream = stream;
+            _buffer = ArrayPool<byte>.Shared.Rent(1024);
```

Comment:
https://github.com/NuGet/NuGet.Client/pull/5529/files#r1415147613
> Did you avoid using an array pool intentionally? I have mixed feelings about whether benchmark code should count the pooled arrays as part of memory usage (BenchmarkDotNet won't count it towards memory usage if the array pool is used). But for production code, I think it's really important to use an array pool - @zivkan 

Nope! Just don't have that habit. I updated it to use the buffer and created an additional function so make sure we return it. Do you think there's a better way to implement the return?

Reply:
System.Text.Json does it via `IDisposable`: https://github.com/dotnet/runtime/blob/2b571c505a444522892a272ea7cf6ff485f59da5/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs#L14

Reply:
ref structs can't implement interfaces but apparently you can just add a public dispose method to the struct and it's fine